### PR TITLE
Menu button column

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

Create menu-button table column.

## 👩‍💻 Implementation

Create `nimble-table-column-menu-button` that creates a menu button in a column and allows slotting a menu into the table to be used with the menu buttons that are created in the column.

This is the first column that hasn't allowed grouping, so I also made a few changes to facilitate not needing to specify a group header view on a column.

## 🧪 Testing

- Manually tested
- Created unit tests for the column
- Created storybook page for the column
- Created matrix tests for the column, including tests with the menu being open

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/table-column/menu-button/cell-view/index.ts
Comment (mollykreis) [inline packages/nimble-components/src/table-column/menu-button/cell-view/index.ts:38]: `tabbableChildren` will need to be implemented here once #2172 is merged.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [inline packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:21]: This `align-content: center` styling does not seem to be having any effect in chromatic, but I'm not sure why. The result is that the text is too close to the top of the menu button rather than being centered. The text alignment within the button looks correct on my machine in chromium, FF, and webkit. I'm open to suggestions of how to resolve the chromatic issue.

**Comment by mollykreis** [comment]
@atmgrifter00, will you buddy this PR for me?

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-opened-matrix.stories.ts
Comment (rajsite) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-opened-matrix.stories.ts:6]: We shouldn't be mixing import path formats, i.e. some directly to paths on disk and some via this format. Lets stick with direct import paths because we would not expect to ship typescript `src` from nimble directly via the package.

Also can you add this format to the block list (`@ni/nimble-components/*` and `@ni/spright-components/*`):

https://github.com/ni/nimble/blob/7450464c52f8dabfcac1f0fbc963a91dc23d4c4d/packages/storybook/.eslintrc.js#L52-L57

Comment (rajsite) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-opened-matrix.stories.ts:6]: We shouldn't be mixing import path formats, i.e. some directly to paths on disk and some via this format. Lets stick with direct import paths because we would not expect to ship typescript `src` from nimble directly via the package.

Also can you add this format to the block list (`@ni/nimble-components/*` and `@ni/spright-components/*`):

https://github.com/ni/nimble/blob/7450464c52f8dabfcac1f0fbc963a91dc23d4c4d/packages/storybook/.eslintrc.js#L52-L57

File: packages/nimble-components/src/table-column/base/cell-view/template.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/base/cell-view/template.ts:39]: CAn we do a prettier disable or something to improve the formatting here?

File: packages/nimble-components/src/table/template.ts
Comment (rajsite) [inline packages/nimble-components/src/table/template.ts:178]: this is pretty hard to understand, maybe a named method helper would be a bit clearer?

File: packages/nimble-components/src/table-column/base/models/column-internals.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table-column/base/models/column-internals.ts:124]: This seems significant to change the defaults of two properties. Is it not? If it is, what are the repercussions? I would presume this would have some amount of cascading implementation, but maybe not?

Comment (mollykreis) [reply packages/nimble-components/src/table-column/base/models/column-internals.ts:124]: Given our mixin pattern, I think it makes sense to disable grouping & sorting by default and only enable those features on a column if the column wants them enabled. The `mixinGroupableColumnAPI` has `groupingDisabled` and `mixinSortableColumnAPI` has `sortingDisabled`, both of which are initialized to `false`, which in turn sets these `ColumnInternals` properties to `false`. Therefore, I don't expect this change to have an impact on existing columns using grouping or sorting; it would just make a column not using those mixins (such as the menu-button column) have those features disabled by default.

File: packages/nimble-components/src/table/template.ts
Comment (mollykreis) [reply packages/nimble-components/src/table/template.ts]: I found a few ways to simplify this statement significantly, but I haven't figured out how to make this a named method helper without losing the observability of things.

Comment (mollykreis) [reply packages/nimble-components/src/table/template.ts]: I talked with @atmgrifter00, and he helped me find a better way to manage this state. It's now part of the `tableData` array, so that simplified things greatly.

File: packages/nimble-components/src/table-column/base/models/column-internals.ts
Comment (rajsite) [inline packages/nimble-components/src/table-column/base/models/column-internals.ts:44]: since this gets initialized in the constructor and is concretely always an array I think we can leave off the optional check

```suggestion
    readonly slotNames: readonly string[];
```

Comment (mollykreis) [reply packages/nimble-components/src/table-column/base/models/column-internals.ts:44]: This is the property on `ColumnInternalsOptions`, not `ColumnInternals`, so it doesn't get initialized in the constructor. Making it optional means that columns that don't care about slots don't need to explicitly set it to be an empty array.

Comment (rajsite) [reply packages/nimble-components/src/table-column/base/models/column-internals.ts:44]: oops, my bad, makes sense

**Review by atmgrifter00 — APPROVED** [review]

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:130]:
```suggestion
                "The name of the slot within the `nimble-table` instance where the menu associated with the column's menu button will be provided",
```
```suggestion
                "The name of the slot within the `nimble-table` instance where the menu associated with the column's menu button will be provided.",
```

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:144]: We don't typically document the default slot for header content on other column types since it's covered by the Table Column Configuration story. Is there any reason to diverge from that pattern for this column?

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:92]: It feels weird to me to open the menu button and select a new value but not have the displayed value update. I suspect that most of our client use cases will want to update the display. If that's true, I propose that:
1. we update this code to call `setData` in response to an item being selected
2. we add documentation saying that clients are responsible for doing this. I think that doc would fit either as a short sentence in the overview section or as a longer section under a `## Usage` heading.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (jattasNI) [inline packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:4]: I was expecting the menu to be the same width as the column. I'm not sure why; maybe because I'm thinking of it like a select? And that would probably not be desirable in all cases, for example if the column is narrow then we wouldn't want the menu to be so narrow that its content is clipped.

I'm curious if we thought about this in the visual design or if we're just getting the default behavior. Assuming it's easy to control the menu width, could you check with Fred/Brandon/the team on what the desired behavior is?

image

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:21]: What do you think about a couple changes to make it easier to experiment with interaction corner cases?
1. add more data so we can see what happens if an open menu is scrolled out of view
2. add an action menu so we can play with keyboard navigation and multiple popups
3. enable selection so we can see how it renders when the row is selected

I'm on the fence; these would more be for me as a tester and might complicate the user-facing documentation / code sample. Maybe I should just pull your branch and try those things myself, but I'm curious if you think others would find them useful.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (jattasNI) [inline packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:1]: What is the state of the visual design right now? Assuming it's not final, do we have an item on Brandon's backlog to provide one? If we think it might be a little while then I guess we could put the docs in the Incubating section initially but I don't feel strongly that's necessary.

Personally I find the bold text seems to overemphasize the column relative to the others.

**Review by jattasNI — COMMENTED** [review]
Just initial storybook and interaction review

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:144]: I didn't realize that it was an established pattern. It just seemed to me that it had been left off by accident on all the columns. I don't mind removing it from here if we think the behavior is documented sufficiently in a different story that applies to all columns.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:4]: This is the default behavior of the menu button itself. I can discuss with Fred and Brandon to see if we're happy with this behavior (both in the column & the stand-alone button).

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:1]: Great question. Last I heard, the visual design was in work. However, that was a few weeks ago. I need to follow up with Fred and Brandon.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:21]: 1. I've added a lot more data to the table (thanks Copilot for generating it all for me 😄)
2. I don't mind adding an action menu, but I wasn't sure how helpful that would be generally for a documentation point of view
3. Similar to the action menu. I don't mind adding selection to the table. I was just trying to be consistent with our other column type stories that don't do anything with selection.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:92]: I've added a `## Usage` heading that mentions that the client needs to handle the `menu-button-column-beforetoggle` event by updating the menu items and handle events on the menu to update the data.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:22]: Wow, Copilot knows the names of Simpsons characters I'd never heard of!

The more I think about 2 and 3, the less I'm able to justify them for clients so I'm ok to leave them off in the name of consistency and simplicity.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (rajsite) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts:238]: In the docs view the menu button isn't able to pop over the content. In combobox / select / menubutton we fix that using this helper:
https://github.com/ni/nimble/blob/2d017ba9169c8d83fb72bf98d6a5f657bd6f6e8d/packages/storybook/src/nimble/combobox/combobox.stories.ts#L106

File: packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts:1245]: I thought when we wrote the keyboard navigation tests we had a vision of parameterizing them so they could easily cover new column types that had interesting keyboard navigation behavior. But I don't remember the details and in a quick scan I don't see the place where we'd expect to do that. @msmithNI do you expect these tests should be changing more significantly for the new column type?

**Review by jattasNI — COMMENTED** [review]
Part way through the implementation and tests

File: packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts
Comment (msmithNI) [reply packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts:1245]: I ended up revamping the tests more based on feedback from Milan. Most of the tests I had for interactive column types were generic, and didn't really need to be run for each additional interactive column type. Those are the tests in this file now (which don't use the anchor column anymore, they use a test column type).

There's a handful of tests in [table-column-anchor.spec.ts](https://github.com/ni/nimble/blob/0f03ba3f181a96a585bb01aaf7da3ce74f6aac22/packages/nimble-components/src/table-column/anchor/tests/table-column-anchor.spec.ts#L650) that cover focusing the control via Tab, and blurring it via Esc, that would probably be good to add for this column type too. Others:
- Checking `tabbableChildren` in various configurations (examples also in the anchor test file)
- Any additional interactions specific to the column type (something with the menu open for this column type, potentially)

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (jattasNI) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts]: I took a stab at addressing this feedback in #2231. If we like that approach and it goes in before this PR, you'll need to update this column's MDX file to refer to the same shared link text.

File: packages/nimble-components/src/table/index.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/index.ts:256]:
```suggestion
    // the uniquified name that the slot should be given.
```

**Review by jattasNI — COMMENTED** [review]
Finished main review, just waiting for open threads to resolve to take another look.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:1]: Good catch! Interestingly, I can't see that behavior with the original colors I had in the example (I guess they aren't long enough strings). But, I see it with the longer colors, so I guess that should be another copilot win since it's the one that generated all of these colors 😄

File: packages/nimble-components/src/table/index.ts
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:256]: Your correction actually isn't correct. Trying to write clear comments with overloaded terms ("slot" in particular) has been difficult, but I'll try to clarify this. What do you think about this?

_Map from the external slot name to the record ID of the row that should have the slot and the unique slot name that the slot should be slotted into._

File: packages/nimble-components/src/table-column/menu-button/tests/table-column-menu-button.spec.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/tests/table-column-menu-button.spec.ts:242]: My current thoughts are that 1 & 2 are overkill since that behavior is completely incapsulated within the menu/menu items. That behavior seems pretty far removed from the table column. But, I'll add a test for 3 to make sure that clicking on an open menu button fires all the expected events and doesn't change table selection.

**Review by rajsite — COMMENTED** [review]
Only partial comments for implementation (got distracted tinkering with the overflow directive 😅)

File: packages/nimble-components/src/table-column/menu-button/index.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/index.ts:85]: Created https://github.com/ni/nimble/issues/2238

File: packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts
Comment (mollykreis) [reply packages/nimble-components/src/table/tests/table-keyboard-navigation.spec.ts:1245]: Thanks for the suggestions. I've added quite a few additional tests following the patterns in the anchor column tests.

File: packages/nimble-components/src/table/index.ts
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:769]: I'm having a hard time understanding why the columnRequestedSlots may contain actionMenuSlots.
Is this a naming conflict guard? If so can that be checked at validation instead?

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table/index.ts
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:769]: Yes, it's to allow the same menu/slot to be used for both a menu-button column and an action menu. We could decide that we didn't want to support that and make the table invalid in that case, but it isn't clear to me why that would be better than allowing a single menu to be used for both. The other alternative to having this specific code is that I could look into rewriting some of the action menu slotting code to reuse some of the new slot tracking code that is being added in this PR.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:4]: Closing this discussion since the current decision is to keep the menu width as-is.

File: packages/nimble-components/src/table/index.ts
Comment (rajsite) [reply packages/nimble-components/src/table/index.ts:769]: > but it isn't clear to me why that would be better than allowing a single menu to be used for both.

yea, fair

> The other alternative to having this specific code is that I could look into rewriting some of the action menu slotting code to reuse some of the new slot tracking code that is being added in this PR.

that does sound nice, I was on the fence about tableData.slots not being used for both or using tableData to hold both (tableData.columnSlots + tableData.actionMenuSlots) maybe as a tech debt and follow-on instead of holding this PR?

File: packages/nimble-components/src/table-column/menu-button/cell-view/index.ts
Comment (msmithNI) [inline packages/nimble-components/src/table-column/menu-button/cell-view/index.ts:39]: Can you doublecheck what this'll do if the table starts out with non-empty string data, then it's replaced with empty strings? 
i.e. [this test](https://github.com/ni/nimble/blob/3165257cb2f224227c4f8dfa348a3dc090feb8ff/packages/nimble-components/src/table-column/menu-button/tests/table-column-menu-button.spec.ts#L109), but first set data as `[{ field: 'value' }]`, then set it to `[{ field: '' }]`

I think in that case, `this.menuButton` will be initialized but not in the DOM, and we don't want `tabbableChildren` to include it in that case.

(That's why the [anchor cell view checks this.showAnchor, not just this.anchor](https://github.com/ni/nimble/blob/297df53de72dc291959ee85bcdff240d96f3b4b3/packages/nimble-components/src/table-column/anchor/cell-view/index.ts#L68))

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:1]: Since it isn't straightforward to make these styling changes to the button, I'll create an issue to tackle this as a follow up task. This way, we can have the necessary discussions around the best way to achieve the desired styling and double check that everything looks as expected with focus/hover states while not blocking the next stages of the column work.

Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts:1]: Created #2247

File: packages/nimble-components/src/table-column/menu-button/cell-view/index.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/index.ts:39]: Good catch! You were right that `this.menuButton` is still initialized (which seems wrong), but I've updated it as you suggested and added a test.

File: packages/nimble-components/src/table/index.ts
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:769]: Created https://github.com/ni/nimble/issues/2248 to track this.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.stories.ts]: Updated to follow the pattern established in your PR.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts:1]: Do you think it would be useful to add this column to Table Keyboard Focus Theme Matrix story? I'm guessing yes since the column has focusable content so it'd be good to validate how its focus state renders. Though we don't currently include anchor columns or action menus in that test so I might be misinterpreting its purpose. @msmithNI feel free to offer an opinion (I know he's OOO today so we can proceed with our best guess).

https://60e89457a987cf003efc0a5b-qjyaqfanzg.chromatic.com/?path=/story/tests-table--table-keyboard-focus-theme-matrix

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.mdx
Comment (jattasNI) [inline packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.mdx:24]: A couple other ideas for docs:
1. at UIWG Fred mentioned this will often be used like a select column. Should we have 1st class docs for this use case? That could mean:
   - updating the example to show a check icon on the currently selected item
   - describing the use case at a high level in the overview
   - describing how to configure the menu (e.g. which check icon to use, which menu item slot to put it in)
2. usage guidance for when to use this column and when to use the action menu

I think both of these are nice-to-have but not required. It's ultimately @fredvisser's call whether they're needed but feel free to do this in a follow up PR if you're unsure of the value and don't want to block this one.

**Review by jattasNI — APPROVED** [review]

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts
Comment (rajsite) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts:1]: I'd think this column should have it's own focus theme matrix test instead of putting all columns in a single test.

Comment (rajsite) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts:1]: Also probably reasonable to address in the style update follow-on

Comment (msmithNI) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts:1]: I don't think it's necessarily needed because the table doesn't add focus visuals to the controls inside cell views, the controls themselves provide that focus styling. Per-column matrix tests could still make sense though I guess.

File: packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts
Comment (mollykreis) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts]: @atmgrifter00 helped me work through the problem. It looks better in chromatic now. @jattasNI - can you double check that it looks correct in safari as well?

Comment (jattasNI) [reply packages/nimble-components/src/table-column/menu-button/cell-view/styles.ts]: Nice! Looks good to me!

image
image

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button-matrix.stories.ts:1]: I'll add a note to #2247 that we should also consider adding a focus theme matrix for the column.

File: packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.mdx
Comment (mollykreis) [reply packages/storybook/src/nimble/table-column/menu-button/table-column-menu-button.mdx:24]: I'll follow up with @fredvisser on what, if any, additional docs we want to add or changes we want to make to the example. For now, I'm resolving the conversation to unblock the PR.

**Pull request merged**
